### ISSUE TITLE
When uploading multiple files, the client does not preserve order

### ISSUE THREAD

**Issue by ghost** [telegramdesktop/tdesktop#27884]
### Steps to reproduce

1. Select multiple files (e.g., mp3 files) in a file browser: `a.mp3`, `b.mp3`, `c.mp3`.
2. Drag and drop into a chat window.

### Expected behaviour

Files are uploaded and displayed in selection order: `a`, `b`, `c`.

This behavior used to be the (undocumented?) status quo until recently, which I personally found useful.

### Actual behaviour

Files are now uploaded in parallel. This is not an issue by itself, and actually noticeably improves upload speeds. During the upload, the order of files looks preserved on the uploading client (until relaunching the client), regardless of when each individual file upload actually finished. However, the order of uploads is not guaranteed anymore, and files may arrive shuffled depending on file size, e.g.: `a`, `c`, `b`.

This change was introduced in a very recent version of Telegram Desktop, most likely in 0dcc439ddae2d686fc4b63e521c1d5eb829bc3ef.

### Operating system

Arch Linux, x86_64, Plasma 6 

### Version of Telegram Desktop

4.16.8

### Installation source

Other (unofficial) source

**Comment by Aokromes** [comment]
it's os "issue" if you select 3 files and drag a file a will be upload 1st, if you select c file c will be upload fist.

**Comment by interruptinuse** [comment]
This is not an OS issue. It's the fact that file uploads are delivered from the server at the time the file is done uploading, and this order used to be deterministic. Now that multiple files are uploaded at the same time, this is no longer guaranteed.

**Comment by Aokromes** [comment]
https://github.com/telegramdesktop/tdesktop/issues/26971#issuecomment-1781162728

**Comment by interruptinuse** [comment]
I'm not using Windows.

**Comment by Aokromes** [comment]
> I'm not using Windows.

still it's the same "bug"

**Comment by john-preston** [comment]
Need to test. I was hoping I did preserve the order.

**Comment by interruptinuse** [comment]
I've been trying to repro this, and so far it's been extremely finicky. I've only triggered the reordering issue with ungrouped MP3 uploads in a channel so far; it seems to work much better after a pause in traffic (subsequent attempts to reproduce are fruitless).

I actually haven't assembled a large enough sample size to confidently say that parallel uploads are to blame, but I don't think I remember it happening before.

Here's what it looks like:

<details>
<summary>1. In the file manager (Dolphin)</summary>

![Screenshot_20240510_144411](https://github.com/telegramdesktop/tdesktop/assets/49772408/40cff4d7-32bd-431a-8d2b-d5843537f5f6)
</details>

<details>
<summary>2. In the Arch Linux 4.6.18 client after uploading (correct order)</summary>

![Screenshot_20240510_145041](https://github.com/telegramdesktop/tdesktop/assets/49772408/615d5d6b-e6b7-4549-8957-1afe88bf0a0d)
</details>

<details>
<summary>3. In mobile and web clients (wrong order)</summary>

![photo_2024-05-10_14-54-44](https://github.com/telegramdesktop/tdesktop/assets/49772408/75afa86e-a1b7-4023-a206-05ce420ad87d)

![Screenshot_20240510_144525](https://github.com/telegramdesktop/tdesktop/assets/49772408/42bd2926-7fb5-4f8d-8893-754a6ff1b498)
</details>

So it does seem to me that it's an ordering heisenbug.

**Comment by interruptinuse** [comment]
Another repro with white noise MP3s.

<details>
<summary>Generator script</summary>

```bash
#!/usr/bin/env bash
set -ex
for i in {01..20} ; do
  COVER_SIZE=500
  head -c "$((3*COVER_SIZE*COVER_SIZE))" /dev/urandom \
    | convert -depth 8 -size "${COVER_SIZE}x${COVER_SIZE}" RGB:- $i.jpg

  DURATION=$(( 100 + ( RANDOM % 100 ) )).$(( RANDOM % 100 ))
  ffmpeg -f lavfi -i "anoisesrc=a=0.1:c=white:d=$DURATION" -i $i.jpg \
    -map 0:0 -map 1:0 -id3v2_version 3 \
    -metadata:s:v title="Album cover" -metadata:s:v comment="Cover (front)" \
    -metadata artist="White Noise" -metadata title="$i" -b:a 192k -c:v copy -y $i.mp3
  rm -v $i.jpg
done
```
</details>

<details>
<summary>Arch client, after uploading</summary>

![Screenshot_20240510_150816](https://github.com/telegramdesktop/tdesktop/assets/49772408/6e0dec27-ac71-461c-8906-6f1c131f3aa4)
</details>

<details>
<summary>iOS client</summary>

![photo_2024-05-10_15-09-52](https://github.com/telegramdesktop/tdesktop/assets/49772408/7f92564d-65b3-430e-9065-84d83052e163)
</details>

**Comment by UpscaleAnon** [comment]
I think it depends on size of file, because I've had this happen a lot in the past with other files. I think it's because telegram processes the files a bit before they get sent in a channel and if the second file is finished processing (not uploading) before the first one, it'll get "sent" first, no matter what you do. It is extremely annoying. Telegram seems more prone to this after the recent update that boosted upload speed (Thank fuck for that update, finally usable speeds).

**Comment by john-preston** [comment]
I couldn't reproduce that.

@interruptinuse Can you please:

1. launch the app
2. enable debug logs by typing "debugmode" in Settings (like a cheat-code in a game)
3. reproduce the bug (upload the files and see them with a wrong order in a mobile app)
4. make screenshots of uploaded files and mobile app files order
5. in Settings type "viewlogs" (like a cheat-code in a game) to see where the logs are stored
6. close the app completely (using Ctrl+Q keyboard shortcut)
7. .zip DebugLogs folder and send it to me at https://t.me/preston (this logs will contain all the network activity from the moment you've enabled them and till quit)

I'll see the send file requests order at least.

**Comment by john-preston** [comment]
I hope this will be fixed in 5.0.2 later today.

**Comment by interruptinuse** [comment]
@john-preston Could not reproduce the issue with `build_5_0_1_1.tar.xz` so far. I did have a few `MSG_WAIT_FAILED` responses in MTP logs, but no reordering issues.

**Comment by interruptinuse** [comment]
Closing as fixed in 84ec2a5f7494ffc2516ecc78ef89aba3da199b33 (5.0.2).

**Comment by interruptinuse** [comment]
Reordering behavior regrettably still appears in 5.1.5 (same platform).
